src/Microsoft.Diagnostics.Runtime/ClrModule.cs

```diff
@@ -31,26 +31,49 @@ public sealed class ClrModule : IClrModule
         private ulong? _size;
         private ClrHeap? _heap;
 
-        internal ClrModuleInfo ModuleInfo { get; }
+        private int initialized;
+        private readonly ulong _moduleAddress;
+        private ClrModuleInfo _moduleInfo;
+
+        internal ClrModuleInfo ModuleInfo
+        {
+            get
+            {
+                if (initialized == 1)
+                {
+                    return _moduleInfo;
+                }
+
+                ClrModuleInfo tmp = _helpers.GetModuleInfo(_moduleAddress);
+
+                if (Interlocked.Exchange(ref initialized, 1) == 0)
```

Comment:
Since the initialization is just assigning the struct you could do a 3-state initialization in the interlocked variable (0=not inialized, 1=initializing, 2=initialization completed) with a CompareExchange transition from 0 to 1, initializing if old state was 0 and spinning while 1. After writing the moduleInfo can Exchange to 2. Not your typical pattern but doesn't seem too problematic?

Reply:
If you don't mind the extra allocation, you could also use `Lazy<T>`.

Reply:
> you could do a 3-state initialization in the interlocked variable

There's no need to do a 3 state initialization.  The issue is that there needs to be a full memory barrier to set state==1 after the assignment.

> If you don't mind the extra allocation, you could also use `Lazy<T>`.

I've avoided Lazy<T> like the plague in ClrMD for years.  It adds so much overhead, but it probably is the right choice if we keep this design.

My main thought is to not have lazy initialization at all in ClrModule, but instead lazily initialize ClrModule objects themselves.  I have to see if that's actually feasible or not though, I haven't dug back through the code yet, but will later this week.

Reply:
Thanks for the comments here. Agreed with @leculver offline to use the Lazy<T> approach, then we can go back writing a lock free solution later if need be.

Under the assumption that `_helpers.GetModuleInfo(_moduleAddress);` always returns the same, the following should be correct.
```C#
get
{
   if (Volatile.Read(ref initialized) == 1)
   {
        return _moduleInfo;
    }
 
    _moduleInfo = _helpers.GetModuleInfo(_moduleAddress);
 
    if (_moduleInfo.Size == 0)
    {
        _size = _moduleInfo.Size;
    }

    Volatile.Write(ref initialized, 1);

    return _moduleInfo;
}
```